Add support for Clang 16

Compiling with Clang 16 shipped with macOS Sequoia (15.0) gives some errors. There are likely more.

```
n file included from src/Ice/generated/Process.cpp:23:
include/Ice/OutgoingAsync.h:359:32: error: declaration shadows a local variable [-Werror,-Wshadow-uncaptured-local]
  359 |             _response = [this, response = std::move(response)](bool ok)
      |                                ^
include/Ice/OutgoingAsync.h:353:36: note: previous declaration is here
  353 |             std::function<void(R)> response,
      |                                    ^
include/Ice/OutgoingAsync.h:395:32: error: declaration shadows a local variable [-Werror,-Wshadow-uncaptured-local]
  395 |             _response = [this, response = std::move(response)](bool ok)
      |                                ^
...
```

See
- https://github.com/llvm/llvm-project/issues/71976
- See https://github.com/llvm/llvm-project/issues/81307

--- a/config/Make.rules.Darwin
+++ b/config/Make.rules.Darwin
@@ -42,7 +42,8 @@ endif
 # On macos, objects are always PIC/PIE, and executables are always PIE. So there is no need to pass -fPIC or -fPIE to
 # the compiler.
 cppflags        = -fvisibility=hidden -Wall -Wextra -Wshadow -Wshadow-all -Wredundant-decls -Wno-shadow-field \
-                  -Wdeprecated -Wstrict-prototypes -Werror -Wconversion -Wdocumentation -pthread \
+                  -Wdeprecated -Wstrict-prototypes -Werror -Wconversion -Wdocumentation -Wno-shadow-uncaptured-local \
+                  -pthread \
                   $(if $(filter yes,$(OPTIMIZE)),-O2 -DNDEBUG,-g)
 
 ifeq ($(MAXWARN),yes)

--- a/cpp/include/Ice/OutgoingAsync.h
+++ b/cpp/include/Ice/OutgoingAsync.h
@@ -23,7 +23,7 @@
 
 #if defined(__clang__)
 #    pragma clang diagnostic push
-// See https://github.com/zeroc-ice/ice/issues/2747
+// See #2747
 #    pragma clang diagnostic ignored "-Wshadow-uncaptured-local"
 #endif
 
@@ -501,7 +501,7 @@ namespace IceInternal
     }
 }
 
-#ifdef __clang__
+#if defined(__clang__)
 #    pragma clang diagnostic pop
 #endif
 

